File: examples/current_manager/current_manager.py
from jormungandr.autodiff import Variable, VariableMatrix
from jormungandr.optimization import OptimizationProblem


class CurrentManager:
    """
    This class computes the optimal current allocation for a list of subsystems
    given a list of their desired currents and current tolerances that determine
    which subsystem gets less current if the current budget is exceeded.
    Subsystems with a smaller tolerance are given higher priority.
    """

    def __init__(self, current_tolerances: list[float], max_current: float):
        """
        Constructs a CurrentManager.

        Parameter ``current_tolerances``:
            The relative current tolerance of each subsystem.

        Parameter ``max_current``:
            The current budget to allocate between subsystems.
        """
        self.__desired_currents = VariableMatrix(len(current_tolerances), 1)
        self.__problem = OptimizationProblem()
        self.__allocated_currents = self.__problem.decision_variable(
            len(current_tolerances)
        )

        # Ensure desired_currents contains initialized Variables
        for row in range(self.__desired_currents.rows()):
            # Don't initialize to 0 or 1, because those will get folded by
            # Sleipnir
            self.__desired_currents[row] = Variable(float("inf"))

        J = 0.0
        current_sum = 0.0
        for i in range(len(current_tolerances)):
            # The weight is 1/tolᵢ² where tolᵢ is the tolerance between the
            # desired and allocated current for subsystem i
            error = self.__desired_currents[i] - self.__allocated_currents[i]
            J += error * error / (current_tolerances[i] * current_tolerances[i])

            current_sum += self.__allocated_currents[i]

            # Currents must be nonnegative
            self.__problem.subject_to(self.__allocated_currents[i] >= 0.0)
        self.__problem.minimize(J)

        # Keep total current below maximum
        self.__problem.subject_to(current_sum <= max_current)

    def calculate(self, desired_currents: list[float]) -> list[float]:
        """
        Returns the optimal current allocation for a list of subsystems given a
        list of their desired currents and current tolerances that determine
        which subsystem gets less current if the current budget is exceeded.
        Subsystems with a smaller tolerance are given higher priority.

        Parameter ``desired_currents``:
            The desired current for each subsystem.

        Raises ``ValueError``:
            if the number of desired currents doesn't equal the number of
            tolerances passed in the constructor.
        """
        if self.__desired_currents.rows() != len(desired_currents):
            raise ValueError(
                "Number of desired currents must equal the number of tolerances passed in the constructor."
            )

        for i in range(len(desired_currents)):
            self.__desired_currents[i].set_value(desired_currents[i])

        self.__problem.solve()

        result = []
        for i in range(len(desired_currents)):
            result.append(max(self.__allocated_currents.value(i), 0.0))

        return result

File: examples/current_manager/src/current_manager.cpp
// Copyright (c) Sleipnir contributors

#include "current_manager.hpp"

#include <algorithm>
#include <limits>
#include <stdexcept>
#include <vector>

CurrentManager::CurrentManager(std::span<const double> current_tolerances,
                               double max_current)
    : m_desired_currents{static_cast<int>(current_tolerances.size()), 1},
      m_allocated_currents{
          m_problem.decision_variable(current_tolerances.size())} {
  // Ensure m_desired_currents contains initialized Variables
  for (int row = 0; row < m_desired_currents.rows(); ++row) {
    // Don't initialize to 0 or 1, because those will get folded by Sleipnir
    m_desired_currents[row] = std::numeric_limits<double>::infinity();
  }

  slp::Variable J = 0.0;
  slp::Variable current_sum = 0.0;
  for (size_t i = 0; i < current_tolerances.size(); ++i) {
    // The weight is 1/tolᵢ² where tolᵢ is the tolerance between the desired
    // and allocated current for subsystem i
    auto error = m_desired_currents[i] - m_allocated_currents[i];
    J += error * error / (current_tolerances[i] * current_tolerances[i]);

    current_sum += m_allocated_currents[i];

    // Currents must be nonnegative
    m_problem.subject_to(m_allocated_currents[i] >= 0.0);
  }
  m_problem.minimize(J);

  // Keep total current below maximum
  m_problem.subject_to(current_sum <= max_current);
}

std::vector<double> CurrentManager::calculate(
    std::span<const double> desired_currents) {
  if (m_desired_currents.rows() != static_cast<int>(desired_currents.size())) {
    throw std::runtime_error(
        "Number of desired currents must equal the number of tolerances "
        "passed in the constructor.");
  }

  for (size_t i = 0; i < desired_currents.size(); ++i) {
    m_desired_currents[i].set_value(desired_currents[i]);
  }

  m_problem.solve();

  std::vector<double> result;
  for (size_t i = 0; i < desired_currents.size(); ++i) {
    result.emplace_back(std::max(m_allocated_currents.value(i), 0.0));
  }

  return result;
}
